### ISSUE TITLE
Support syntax highlighting with tree-sitter

### ISSUE THREAD

**Issue by fcurts** [Microsoft/vscode#50140]
Please consider supporting [tree-sitter](https://github.com/tree-sitter/tree-sitter) grammars in addition to TextMate grammars. TextMate grammars are incredibly difficult to author and maintain and impossible to get right. The over 500 (!) issues reported against https://github.com/Microsoft/TypeScript-TmLanguage are a living proof of this.

This presentation explains the motivation and goals for tree-sitter: https://www.youtube.com/watch?v=a1rC79DHpmY

tree-sitter already ships with Atom and is also used on github.com.

**Comment by Kroc** [comment]
This may help in the future with the whole 'embedding one language in another', which is an enfant terrible when it comes to TextMate grammars.

**Comment by omniomi** [comment]
There's also a request in #5408 for .sublime-syntax which has been open since Apr 2016 which would also be a step up from .tmLanguage.

While tree-sitter has an awesome concept I can't say the idea of writing grammar in JavaScript is all that appealing.

**Comment by fcurts** [comment]
@omniomi tree-sitter also supports writing grammars in pure JSON if that's what you prefer. The main & dramatic advantage of tree-sitter is that it's a full parsing system and not an ad-hoc, underspecified, horrifyingly complex yet extremely limited regex contraption.

**Comment by ahuertabhg** [comment]
Integrating tree-sitter would help solve this issue https://github.com/OmniSharp/omnisharp-vscode/issues/2461

**Comment by sean-mcmanus** [comment]
@aeschli Atom has switched to tree sitter for C++ and no longer fixing issues with Text Mate: https://github.com/atom/language-c/issues/232#issuecomment-426018195 . Please advise on how we should proceed for improving the C++ syntax highlighting/etc. experience.

**Comment by maxbrunsfeld** [comment]
:wave: Just to reiterate - the Atom team doesn't intend to disrupt other apps like VSCode that are using modules like `language-c`. We will definitely continue to accept good PRs that update the text-mate grammar.

The reason that we've been closing issues like that is just to be explicit about the fact that our team won't be prioritizing work on them in the future, since Atom is moving away from text-mate grammars.

**Comment by beeuhtricks** [comment]
Tree-sitter is extensible for other programming languages, and in particular already supports Rust and Ruby as well. Are the Visual Studio APIs ready to be extended with new language support in those ways?

**Comment by Geobert** [comment]
I'm wondering if tree-sitter can solve this https://github.com/Microsoft/vscode/issues/51157

**Comment by Stanzilla** [comment]
@bobbrow is that a finite decision? Would have been nice to share the code with Atom here.

**Comment by Astrantia** [comment]
No plans for this in 2018?

**Comment by github-yxb** [comment]
It's going to be 2019!!

**Comment by fcurts** [comment]
> I fully agree with you that TextMate grammars are challenging to implement and have limitations, but it's always a lot of work to create and maintain a grammar. That will not be different with Tree-Sitter.

@aeschli I meanwhile re-implemented my TextMate grammar with tree-sitter because the former proved unmaintainable (templated regexes up to 400 characters long, etc.). Developing the tree-sitter grammar and highlighter from scratch took three days, compared to three weeks for the TextMate grammar. The new highlighter works better and is easier to maintain.

**Comment by jeff-hykin** [comment]
@fcurts Could I see your re-implementation? I currently maintain the TextMate C/C++ grammar. Its actually really easy to maintain now that its written in Ruby with actual variables and functions instead of Regex, but it doesn't fix the inherent limitations of the TextMate engine.

I'd love to figure out how to convert it to a tree sitter syntax, the Atom tree sitter C++ syntax has problems. To be honest even after spending awhile looking at it, I have no idea how to use it or how it works. Its missing functionality like lambda support, compile attributes, templated function calls, and macro calls. I went and looked at the source code [here](https://github.com/tree-sitter/tree-sitter-cpp/blob/master/src/grammar.json), but its pretty unintuitive and its >6000 lines long. Which is double the size of the TextMate grammar.

**Comment by maxbrunsfeld** [comment]
> I went and looked at the source code here, but its pretty unintuitive and its >6000 lines long. Which is double the size of the TextMate grammar.

No, the source for the Tree-sitter C++ grammar is [here](https://github.com/tree-sitter/tree-sitter-cpp/blob/master/grammar.js). It's 669 lines of JavaScript (not counting the C grammar, which it inherits from).

> Its missing functionality like lambda support, compile attributes, templated function calls, and macro calls.

It does fully support lambdas and template function calls, as far as I know. You can see the test suite for these features [here](https://github.com/tree-sitter/tree-sitter-cpp/blob/master/corpus/expressions.txt#L154). I do think that compile attributes are unimplemented.

**Comment by jeff-hykin** [comment]
That's great to hear, and thanks for correcting me! I'll try looking into that file to get a better understanding.

In Atom lambdas are still partially incorrectly marked, although it isn't visually messed up. Just things like the `->` is marked as a member-access operator. Templated function calls are also not colored like functions (they have no color / theme-scopes). This could be something specific to Atom and not the grammar though.

**Comment by dberlin** [comment]
FWIW: To make this more annoying for folks to choose - i have patches to add incremental parsing/lexing support to ANTLR4 (both the main java runtime and the optimized typescript runtime port).

Incremental parsing is submitted to both repositories at this point, incremental lexing is not but will be in the next week or two.

I already use the incremental parser/lexers in my vscode extensions.

The lexing is the same set of algorithms tree-sitter uses, the incremental parsing is actually simpler because LL is top-down.

Speed wise, it can also relex/reparse on every keystroke with no issue.
I mention it since ANTLR has a large collection of language grammars as well.

**Comment by chfritz** [comment]
@aeschli another benefit of tree-sitter besides syntax highlighting is that it lends itself to a much better indentation logic (https://github.com/atom/atom/pull/18321). Just take a look at [this example file](https://github.com/chfritz/atom-sane-indentation/blob/master/spec/fixtures/sample.js) and compare the indentation there with what VSCode will currently produce. For example:

With tree-sitter :
```js
foo( 2,
  {
    sd,
    sdf
  });

foo( 2, {
  sd,
  sdf
});
```

vscode:
```js
foo( 2,
  {
    sd,
    sdf
  });
  
  foo( 2, {
    sd,
    sdf
  });
```
The second call to `foo` is at the root level, so why is it indented? The answer is quite simple: an inductive indentation approach that just considers the previous line to determine the indentation for the current line cannot handle multiple scopes opening on the same line but closing on different ones, which is what these examples show. There are ways to deal with that, but they are not as flexible to make the indentation look the way people expect them to be.

If you add tree-sitter to vscode then I'm happy to try and port https://github.com/atom/atom/pull/18321 to vscode.

**Comment by georgewfraser** [comment]
I recently made an extension that adds support for tree-sitter by replacing the builtin grammar with a simplified grammar that just colors literals and keywords, and then using the setDecorations API to apply tree-sitter based coloring to the tricky parts:

https://marketplace.visualstudio.com/items?itemName=georgewfraser.vscode-tree-sitter

For example, this is what Go looks like before and after installing the extension:

<img width="921" alt="Screen Shot 2019-05-19 at 12 00 34 PM" src="https://user-images.githubusercontent.com/1369240/57987265-f7279d00-7a33-11e9-9aea-43964d334e35.png">

It currently supports 
- Go
- C++
- Typescript

and it's straightforward to add any of the available [tree-sitter languages](http://tree-sitter.github.io/tree-sitter/).

**Comment by Razzeee** [comment]
Thanks for doing this, I really like the idea. I looked at doing that for elm and decided against it for the moment. I think it would be better to have it in my elm plugin and the best approach would be to actually have it in the language server, whenever this gets merged. https://github.com/Microsoft/language-server-protocol/issues/513

**Comment by Razzeee** [comment]
@georgewfraser does this approach also work for the embedded markdown code? For e.g. when you get a completion and it shows some docs, which have some code embedded?

**Comment by dannymcgee** [comment]
> While tree-sitter has an awesome concept I can't say the idea of writing grammar in JavaScript is all that appealing.

I know this is an old comment, but I don't really understand this logic. I've written (or attempted to write) several personal-use grammars for VS Code, and everything about the tmLanguage.json system is hopelessly awkward and unintuitive.

1. There's a steep learning curve to get even a basic grasp on how [TextMate grammars are structured](https://macromates.com/manual/en/language_grammars)
2. It uses Oniguruma for its regex library, making the learning curve even steeper
3. Since the regexes need to be written as strings, extra escape characters are needed, making them more difficult to read and write
4. JavaScript is the single [most widely used language on GitHub](https://octoverse.github.com/projects.html#languages), and writing a grammar for Tree Sitter is [relatively intuitive](https://github.com/tree-sitter/tree-sitter/blob/master/docs/section-3-creating-parsers.md#writing-the-grammar), so not only is it a vastly more powerful solution, but the barrier to entry for potential contributors is much lower

I realize that [semantic highlighting](https://github.com/microsoft/vscode/issues/585) is on VS Code's roadmap so this may be a moot point entirely (although frankly it has been "on the roadmap" for years with no movement), but I am really not seeing any downsides to "writing grammar in JavaScript" vs. VS Code's current implementation, which is frankly a nightmare

**Comment by jeff-hykin** [comment]
I think what @georgewfraser has made is great in the sense that; I don't think we need to wait on the VS Code's Core team to start work on a tree parser. I mean, sure, it being merely an extension isn't amazing, but extensions can do almost everything core does, which is more than enough to kickstart the work show that it can integrate well with VS Code. The more support we build for it the faster it will get merged into core, and I applaud @georgewfraser for effectively taking the first step 👏👏👏

**Comment by EvgeniyPeshkov** [comment]
Hello everyone. I've developed and published syntax highlighting extension based on Tree-Sitter. It provides universal syntax coloring engine for almost any programming language (currently, C and C++ are supported OOTB).
It's very easy to add support for a new language. I'm planning to write HowTo in the next couple of days, but you can figure it out from the source code, it's very simple and straightforward. Contributions are welcome.
I've been using the extension by myself for a month, so I suppose it's ready for public use. At least it can be useful until VSCode core provides stronger syntax parser.

[You can install it from VSCode Marketplace.](https://marketplace.visualstudio.com/items?itemName=evgeniypeshkov.syntax-highlighter)
[Or download .vsix package from GitHub page and install it manually.](https://github.com/EvgeniyPeshkov/syntax-highlighter/releases)
[**Please note**, that extension published in VS Code Marketplace will only work in Windows-x64](https://github.com/EvgeniyPeshkov/syntax-highlighter#distribution).
[For other operating systems, please download pre-compiled .vsix package.](https://github.com/EvgeniyPeshkov/syntax-highlighter/releases)
This will be fixed in the near future with one of the next updates.
Alternatively, you can [build extension from sources](https://github.com/EvgeniyPeshkov/syntax-highlighter/blob/master/BUILD.md).

I've noticed, that @georgewfraser published his implementation a couple of days ago. I suppose we had the same thoughts. I'm very glad that qualitative alternatives to limited TextMate grammars begin to appear. Thank you, @georgewfraser.

**Comment by Stanzilla** [comment]
You guys should join forces :)

**Comment by jeff-hykin** [comment]
As an update to anyone interested in these extensions:
Manual .visx installation is no longer required. @georgewfraser, @EvgeniyPeshkov, and I got both of the extensions running with Web Assembly so now they work out of the box on basically any platform.

**Comment by Paul-Sinclair** [comment]
Does that extension fix this issue https://github.com/OmniSharp/omnisharp-vscode/issues/2461 ?

**Comment by Razzeee** [comment]
It probably would, but the tree-sitter implementation for csharp is incomplete and i think it doesn't know what a define is right now.
https://github.com/tree-sitter/tree-sitter-c-sharp

**Comment by Geobert** [comment]
Even if these extensions provide better syntax highlighting, it not ideal because we can't use them with other extensions that uses the same Decoration API (see https://github.com/microsoft/vscode/issues/74692)

**Comment by Paul-Sinclair** [comment]
Well that's disappointing, VScode wouldn't be bad for C# development if MS would properly support the language features. Will continue to use Visual Studio for C# projects. Shame really as VScode would work much faster for Unity projects and if they improved the debugging features on C# would make it the ideal replacement VS.

I'm just surprised it hasn't been more of an issue.

**Comment by jeff-hykin** [comment]
To continue making progress on this, we can use the repo [here](https://github.com/jeff-hykin/experimental-tree-sitter/issues) to track and discuss the separate issues blocking the tree sitter implementation. I've added issues/labels for all of the known problems including the one @Geobert just pointed out. Feel free to ask questions, make feature requests, and ask for time estimates, or subscribe if you want incremental updates. I'll keep the readme updated with the general progress/timeline of the different extension implementations.

That way the extension contributors can have a central place for tracking/fixing issues, and this thread won't become bloated with every possible topic/question related to the tree-sitter.

If there are any major breakthroughs to any of the extensions (or core), we'll make sure post on this thread. Right now the largest challenge is getting generic long term support for themes, with the secondary issue of fast colorization. It will likely be a month before the next major step forward.

**Comment by atomiks** [comment]
## Incomplete list of features of a perfect syntax highlighter for me

Aside from the obvious like `string`, `comment`, `keyword`, there needs to be a level of granularity for me, and of course semantic recognition.

These are all in the context of JavaScript specifically.

- It can distinguish types of function contexts: `declaration`, `call`, `method-call`

> These are possible in TextMate grammar, but `method-call` isn't added yet

```js
// declaration
function func() {}
// call
func();
// method-call
obj.func();
```

- It scopes parameter variables within a function body

> Requires semantic highlighting

```js
let outer = 0;

function func(a, b) {
  let inner = 1;
  // `a` and `b` should have a special parameter scope, both in the def 
  // and when using them
  // `inner` and `outer` are plain variables
  return a + b + inner - outer;
}
```

- You can specify particular words to have their own scope, e.g. built-ins

> Possible in TextMate, but it was incomplete and was recently removed, I'd like to manually define these if possible

```js
console;
window;
document;
this;
setTimeout;
requestAnimationFrame;
arr.slice();
document.querySelector();
```

- It can distinguish data structure syntax

> Requires semantic highlighting

```js
const objectLiteral = {}; // {} should not have a `punctuation` scope
const arrayLiteral = []; // [] should not have a `punctuation` scope

// But they should have `punctuation` scope here
{ /*  inside a block */ }
array[0] = 2; 
```

- It can distinguish levels of object property access and objects

> I think this is possible in TextMate, I believe this feature may exist already, but sometimes was buggy due to built-in DOM scopes/names

```js
// `obj` should have a "top-level" object scope
// `one` and `two` should have a "sub-object" scope
// `three` should have a last-property scope
obj.one.two.three;

// Also object literal def properties have their own scope
obj = {
  prop: true // `prop` has its own scope
};
```

- It can distinguish the type of a variable if it's static

> Requires semantic highlighting

```js
class Hello {}
Hello; // should be same as the class def

// Primitives constants can (optionally) have their related scope
const number = 0;
number; // should have `number` scope along with regular `variable / constant` scope

// Constants also scoped when declaring & using them
const CONSTANT = '';
```

- Certain keywords can have their own scopes

> Possible in TextMate, but not every keyword has its own scope and it gets grouped with other ones in the current language grammar

```js
// `import` and `from` should have a particular module keyword scope
import {module} from 'pkg';

// `function` and `return` should have their own scopes if they want as well
function x() {
  return true;
}
```

---

The current TextMate implementation already has a lot of these nice features, and when Atom switched to their new tree-sitter syntax, it lost parts of these features. So while it gained nice features, it also lost some. 

Many people don't realize these particular granular features can be important (or realize they should exist), but they are important to me. So whatever new implementation gets added, please make sure these are all possible in their respective languages 👌

**Comment by flowchartsman** [comment]
As mentioned by @sean-mcmanus  [here](https://github.com/microsoft/vscode/issues/50140#issuecomment-426028148) and [here](https://github.com/microsoft/vscode/issues/50140#issuecomment-452141813), the dependency on the now unmaintained Atom TextMate grammars, is preventing issues from being addressed in the display of C/C++.  You can [add Go to the list now, too](https://github.com/atom/language-go/pull/178). The new `%w` format verb is not highlighted, and even the simple change necessary to make this possible can't go in because neither side is willing to make the change. This is only going to get worse.

**Comment by matter123** [comment]
@floriantraber The C/C++ grammar is now being maintained at https://github.com/jeff-hykin/cpp-textmate-grammar/ if you have an issue with the syntax highlighting, please file it there.

**Comment by flowchartsman** [comment]
Please read messages before responding. I was talking about Go.

**Comment by flowchartsman** [comment]
I am not talking about C/C++, I'm talking about Go and, presumably any other language for which the highlight grammar is generated from an unmaintained source.

**Comment by jeff-hykin** [comment]
@flowchartsman clone the atom repo, and add the fix to it. That's the beauty of open source code. You can even make an extension that instantly applies the fix so that you don't have to wait for official VS Code support to start seeing the benefits. I created a repo [here](https://github.com/jeff-hykin/better-go-syntax) fully setup for publishing the go syntax as an example.

If there is a better maintained version of TextMate for go lang, I'm confident the VS Code team will switch to it. Someone just needs to take the initiative to create or suggest that more-maintained repo. @matter123 and I did it for C++, and we're working on tools to make it easier for others to do the same. I don't think Atom or VS Code teams are being stubborn sticking to frozen code.

**Comment by sean-mcmanus** [comment]
There's a pull request at the Go extension for adding tree sitter support, but it's been sitting around since June: https://github.com/microsoft/vscode-go/pull/2555 . @ramya-rao-a What's the plan for improving syntax colorization for Go?

**Comment by flowchartsman** [comment]
> I don't think Atom or VS Code teams are being stubborn sticking to frozen code. Someone just needs to get their hands dirty and implement/publish the fix themself.
@jeff-hykin that kind of kicks the can on the main thrust of the issue here (multi-language dependence on stale atom code and difficult-to-maintain grammars), but I'll take it for now. Seems downright neglectful if I don't fork/PR now. I'll leave this issue to proceed on its own for now.

**Comment by flowchartsman** [comment]
as per https://github.com/microsoft/vscode/issues/50140#issuecomment-532941563

**Comment by fcurts** [comment]
Any updates?

We'd love to support VSCode again, but bringing back our TextMate grammar is not an option. Our tree-sitter grammar has proven to be far easier to maintain, and the highlighting is much better (in Atom).

**Comment by Stanzilla** [comment]
@fcurts https://github.com/microsoft/vscode/issues/77140

**Comment by kolya-ay** [comment]
If VSCode team don't want to rely on Atom's original C-based tree-sitter implementation (It's easy to imagine many different reasons not to use native node modules in the core), there might be worth to consider [Lezer](https://lezer.codemirror.net/) an JS-based implementation of the idea [made by Marijn Haverbeke](https://marijnhaverbeke.nl/blog/lezer.html). The authorship almost guaranties great code quality and maintainability. As a benefit VSCode can share grammars with upcoming CodeMirror version which, without any doubt, will be written for any possible language.

**Comment by Razzeee** [comment]
There are also docs for the new (reworked) tree sitter highlighting now https://tree-sitter.github.io/tree-sitter/syntax-highlighting

**Comment by adrijshikhar** [comment]
Hey everyone, I am a newbie to all this VS Code extension stuff. I wanted to learn how syntax highlighting takes place in the VS Code extension. I started working on a personal project to use TextMate grammar for highlighting javascript in order to learn. Over time, it grew complex. From this thread, I came across a new concept of tree-sitter. So can anybody help me or suggest me nice documentation on, how to port my existing syntax highlighting extension from TextMate grammar to tree-sitter, or if anyone has implemented tree-sitter syntax highlighting for javascript, please reply to this. I can really use some help. Thanks.

**Comment by michaelblyons** [comment]
>  or if anyone has implemented tree-sitter syntax highlighting for javascript, please reply to this.

You seem like a nice guy, so I'll try not to make this too harsh: [The homepage](https://tree-sitter.github.io/tree-sitter/) of Tree Sitter has a link to [the implementation of JavaScript](https://github.com/tree-sitter/tree-sitter-javascript) by TS's original creator. Good luck, and maybe hunt a little harder next time you have a question like this one.

**Comment by jeff-hykin** [comment]
@adrijshikhar This isn't really the right thread for this, a better place would be on the tree sitter repo [here](https://github.com/tree-sitter/tree-sitter) 

What is relevant here is that: if you port your syntax from TM to the Tree Sitter it won't be natively supported in VS Code and the non-native support is still very rough. The Tree Sitter is awesome, so I highly recommend learning it. However if you want something working in VS Code, me and another guy have been working on a library for making maintainable TextMate grammars for more than a year now. The library and documentation was finished recently, and I've been working on creating a tutorial. I'll likely publish the library and the tutorial in the next month on Medium under "Make a TextMate grammar (without wanting to kill yourself)"

**Comment by DrSensor** [comment]
Has anyone in the vscode team explored using [syntect](https://github.com/trishume/syntect) engine?
Converting TextMate grammar into sublime syntax definition is pretty straightforward.

**Comment by jeff-hykin** [comment]
The conversion is straightforward, but the Sublime engine still has the vast majority of the same problems.

**Comment by fcurts** [comment]
2 years later, is there any consensus on how and when to offer a better alternative to TextMate grammars?

**Comment by dannymcgee** [comment]
> 2 years later, is there any consensus on how and when to offer a better alternative to TextMate grammars?

@fcurts "Semantic highlighting" has been available for a short while and it's basically the best syntax highlighting possible, especially when combined with the TM grammars.

It's only available for a few languages so far (It might still be in preview? Don't quote me on that), but bassically, it leverages a language service to determine the actual semantic type of any identifiers in your code and highlights them according to your theme's rules for that type. For tokens that don't have or don't need semantic information from a language service (like keywords and punctuation), it falls back to the TM grammars.

For example, in JavaScript:

```javascript
const onClick = (event) => {
  // do something
}

document
  .querySelector('.my-button')
  .addEventListener('click', onClick) // <-- `onClick` gets colored as a function thanks to semantic highlighting
```

You need to be using a theme that has semantic highlighting enabled (the default themes do), or manually set it to enabled via a user setting override. As far as I'm aware, the supported languages currently are JavaScript, TypeScript, C++ (via the C++ extension), and C# (via the C# extension). There may be more. It's up to the author of any given language extension to provide the semantic highlighting tokens.

**Comment by Razzeee** [comment]
F# also supports it.

But the language server spec for this is still preview and subject to change, as far as I followed it.

**Comment by MewX** [comment]
Currently there's no available tree-sitter option unfortunately...

The current extension is broken due to its wasm-based implementation: https://github.com/georgewfraser/vscode-tree-sitter/issues/28

Btw, Atom has supported tree sitter 2 years ago: https://github.blog/2018-10-31-atoms-new-parsing-system/

**Comment by dannymcgee** [comment]
> Btw, Atom has supported tree sitter 2 years ago: https://github.blog/2018-10-31-atoms-new-parsing-system/

The Atom team developed Tree-Sitter for use with Atom, so it's not really surprising that it's the first (only?) editor to have adopted it. It literally says that in the second paragraph of your link.

VS Code's new [Semantic Highlighting](https://github.com/microsoft/vscode/wiki/Semantic-Highlighting-Overview) implementation is now more accurate than Tree-Sitter since it uses a language service that analyzes your entire project (Tree-Sitter AFAIK has no information about the syntax tree outside of the file it's currently highlighting.)

**Comment by jeff-hykin** [comment]
@dannymcgee semantic highlighting wouldn't replace Textmate. The tree sitter replaces textmate, and semantic highlighting would stay as-is.

**Comment by Khady** [comment]
Having tree sitter and providing an api for it would open the door for many interesting extensions too. We can imagine a paredit that is semantically correct for example.

> it's not really surprising that it's the first (only?) editor to have adopted it. It literally says that in the second paragraph of your link.

neovim has it too.

**Comment by XeonG** [comment]
Has this issue https://github.com/OmniSharp/omnisharp-vscode/issues/2461 been fixed yet? or is vscode still inferior to visual studio for better c# development

**Comment by tristan957** [comment]
This issue is finally on the second page of the issues backlog which is nice to see. Visibility on this issue is probably pretty low just because most end users don't know how painful Tm grammars are until they run into syntax highlighting issues. @jeff-hykin maybe it would help get more visibility if you advertised it on the "Better C/C++ Syntax" readme?

**Comment by XeonG** [comment]
Hopefully fixed soon.. Unity game dev is shit with Vscode for a number of reasons.. one of the biggest being so much of the code will use #if #else #end wrappers for different plugins, dev builds, etc  any having all that unused code not visually greyed out is just rubbish to work with.. might aswel use notepad.

**Comment by jeff-hykin** [comment]
> end users don't know how painful Tm grammars are until they run into syntax highlighting issues. @jeff-hykin maybe it would help get more visibility if you advertised it on the "Better C/C++ Syntax" readme?

@tristan957 

With 330 issues (230 closed) on what is essentially a static json file, some of which are unsolvable, a general explanation in the readme is not a bad idea.

**Comment by tristan957** [comment]
@jeff-hykin since you are familiar with at least enough of VSCode to put together syntax definitions, is it possible to have tree-sitter integration in an external extension? I think as Neovim 0.5 has been developed, the tree sitter integration actually lives in a separate extension, at least for now. https://github.com/nvim-treesitter

**Comment by jeff-hykin** [comment]
@tristan957 
https://github.com/jeff-hykin/experimental-tree-sitter
https://github.com/georgewfraser/vscode-tree-sitter (<- mine is just a generic-ifyed a fork of his)

1. The problem is the extension uses decorations, which are slow. Really Really Really slow
2. There is no good way of accessing the user's theme settings. George's extension has its own colors independent of whatever your theme is. (You can manually customize the colors)
3. Assuming it did have access (there are hacky ways to gain access), there needs to be an implementation of a converter from the Textmate theme to the Tree sitter query language.

And there's smaller problems along with those. #1 is by far the most difficult issue, and the semantic highlighting might fix that, I'd have to learn more about the API.

I imagine Neovim gives extensions a lot more control than VS Code does with it's extensions.

**Comment by NullVoxPopuli** [comment]
Any updates on this? Tree-sitter's nested language parsing has been amazing in neovim.

LSP enhanced highlighting is great and all, but it's still just per language... Multi language support would still be super annoying. 
With tree sitter, we could have not only proper syntax highlighting for *any* combination of languages in the same file, but we could use the same language switching logic to run multiple language servers on the same file.

Would be great of vscode could catch up to the terminal-based editor. ;)

**Comment by sfc-gh-pkommini** [comment]
Any chance this is a go?

**Comment by altosaar** [comment]
Any updates on this? Would help greatly for accessibility and voice input with VS Code :(

**Comment by XeonG** [comment]
Any year now Vscode....

**Comment by ThePrimeagen** [comment]
> Any year now Vscode....

Yes. Tree sitter is used in tons of actors now. Vim and emacs support it.

Writing contextual plugins is incredibly simple too

**Comment by parascent** [comment]
This I feel is a must have for all modern text editors along with LSP.

**Comment by o314** [comment]
Assume your ambition Microsoft. You want to be an opensource leader, you want to be the greatest ide ?
So use the correct technological stack at the parsing level for that.
It will not be done with regex, and reexplaining it in the 202x is a fool's game

**Comment by SNDST00M** [comment]
There should be a way to get a LSP running a tree-sitter configuration.

**Comment by SNDST00M** [comment]
I have found the most probable issue with the tree-sitter extension and I hope a extended support fix is released so that we can use tree-sitter in Visual Studio Code

**Comment by dannymcgee** [comment]
> There should be a way to get a LSP running a tree-sitter configuration.

Just spitballing, but it should actually be feasible to make sort of a "general purpose" LSP extension that takes a map of VS Code language IDs to tree-sitter parsers as configuration, and then uses those parsers to provide semantic tokens for the active file if it matches one of the supported languages. I'm not sure how this would work in case of a conflict though (i.e., if you have two extensions trying to provide tokens for the same language).

But for extension authors wanting a way to add support for some language without having to mess with TextMate grammars, that is absolutely an option. You don't even need to use the LSP protocol for that, you could just make a "standard" VS Code language extension and use a tree-sitter parser as the engine for everything — document symbols, semantic tokens, hovers, whatever.

**Comment by SNDST00M** [comment]
The extension by @jeff-hykin and @georgewfraser doesn't use LSP yet, but I'd like it to. Heck, I'd learn the API for LSP if it meant I could port tree-sitter without those slow decorators.

**Comment by lnicola** [comment]
> I'd learn the API for LSP if it meant I could port tree-sitter without those slow decorators.

You mean this? https://microsoft.github.io/language-server-protocol/specification#textDocument_semanticTokens

**Comment by Razzeee** [comment]
> Assume your ambition Microsoft. You want to be an opensource leader, you want to be the greatest ide ?
> So use the correct technological stack at the parsing level for that.
> It will not be done with regex, and reexplaining it in the 202x is a fool's game

Yeah, if you read up on this, you can find out, that they want to use the parser that's at the core of visual studio. So they're aware, that regex isn't the way to go.



> The extension by @georgewfraser doesn't use LSP yet, but I'd like it to. Heck, I'd learn the API for LSP if it meant I could port tree-sitter without those slow decorators.

What are you trying to do? I actually think it's a bad idea, to have extensions sliced like that (all languages, one feature) instead of the other way around.

**Comment by Milo123459** [comment]
My friend was talking about this issue and I had an "idea" - couldn't someone make a plugin to convert TreeSitter grammars to TextMate? It'd mean you'd be able to use TreeSitter without VSCode needing to change anything

**Comment by SNDST00M** [comment]
> What are you trying to do? I actually think it's a bad idea, to have extensions sliced like that (all languages, one feature) instead of the other way around. It might make more sense to go for something like an api helper, see the old testing extensions.

I agree, but if staff aren't budging on this, `vscode-tree-sitter` is also available as an NPM module. Thankfully the deprecation hasn't been pushed through on NPM or elsewhere.

**Comment by SNDST00M** [comment]
> My friend was talking about this issue and I had an "idea" - couldn't someone make a plugin to convert TreeSitter grammars to TextMate? It'd mean you'd be able to use TreeSitter without VSCode needing to change anything

Unfortunately, Tree-sitter offers a versatile parser that can't be expressed fully as a regular expression. Tree-sitter to Textmate is usually not as bad as parsing HTML with regular expressions, but it's a rather heavy downgrade, especially as Textmate doesn't support variable length lookbehinds.

**Comment by Milo123459** [comment]
Oh that makes sense

**Comment by dannymcgee** [comment]
> Heck, I'd learn the API for LSP if it meant I could port tree-sitter without those slow decorators.

It's honestly a lot more straightforward than you might think, just a matter of trudging through Microsoft's docs and example repos. And like I said, you don't even need to use a language server since tree-sitter has JavaScript bindings.

Here's a really straightforward example from a project of mine that I never finished:

```typescript
// extension entry point
import { ExtensionContext, languages } from "vscode";
import SemanticTokensProvider, { SEMANTIC_TOKENS_LEGEND } from "./semantic-tokens";

export function activate(ctx: ExtensionContext) {
   let provider = languages.registerDocumentSemanticTokensProvider(
      { language: "glsl" },
      new SemanticTokensProvider(),
      SEMANTIC_TOKENS_LEGEND,
   );
   ctx.subscriptions.push(provider);
}
```
```typescript
// semantic-tokens.ts
import {
   DocumentSemanticTokensProvider,
   SemanticTokens,
   SemanticTokensBuilder,
   SemanticTokensLegend,
   TextDocument,
} from "vscode";

import lexer, { TokenType } from "./lexer";
import parser from "./parser";

export enum SemanticToken {
   Macro = "macro",
   Function = "function",
   Param = "parameter",
   Variable = "variable",
}

export const SEMANTIC_TOKENS_LEGEND = new SemanticTokensLegend([
   SemanticToken.Macro,
   SemanticToken.Function,
   SemanticToken.Param,
   SemanticToken.Variable,
]);

export default class SemanticTokensProvider implements DocumentSemanticTokensProvider {
   provideDocumentSemanticTokens(doc: TextDocument): SemanticTokens {
      let builder = new SemanticTokensBuilder(SEMANTIC_TOKENS_LEGEND);
      let tokens = lexer.tokenize(doc);

      tokens
         .filter(tok => tok.type === TokenType.Ident)
         .forEach(tok => {
            let decl = parser
               .getScopeAt(doc, tok.range)
               .findDeclOf(tok.data);

            if (!decl) return;

            builder.push(tok.range, decl.semanticType);
         });

      return builder.build();
   }
}
```

You would just need to replace the custom lexer/parser there with tree-sitter, but you get the idea. I only processed the identifiers because I prefer leaning on TextMate for the unambiguous stuff like keywords and literals, but you could highlight everything with semantic tokens if you wanted. You would probably also want to look into how to take advantage of tree-sitter's incremental parsing for better performance.

**Comment by SNDST00M** [comment]
I am actually working on a LSP for MATLAB based on `vscode-textmate`. This looks pretty similar!

**Comment by Razzeee** [comment]
> And like I said, you don't even need to use a language server since tree-sitter has JavaScript bindings.

Unless you want to support other editors too.

**Comment by jeff-hykin** [comment]
Having a tree-sitter-to-LSP tool would absolutely be the best next step forward. It doesn't solve all of the problems, but it solves the biggest one.


@SNDST00M (and others) if you're interested
 in messing with the code to take advantage of the LSP let me know and I'll get you access to the [experimental tree sitter](https://github.com/jeff-hykin/experimental-tree-sitter). 

Integrating the LSP likely wouldn't be too hard, however there is a sperate challenge. The tree sitter grammars are actually too powerful for the theming system. The grammars tell you everything about every token (which is awesome) but  themes are way more dumb about what they color. There isn't a good generic way to map tree-sitter-information to the existing themes. Either all existing themes would break (no backward compatiblity), or there needs to be a custom converter for each language, in addition to a language parser.


Once those values have been converted, you can then likely send them to the LSP. Right now that conversion is being manually done inside of the extension, more as proof of concept than anything else.

**Comment by ThePrimeagen** [comment]
Does Microsoft have a license? https://regexlicensing.org/ I would prioritize this before vscode gets cited for use of a regex without a license.

**Comment by dannymcgee** [comment]
> > And like I said, you don't even need to use a language server since tree-sitter has JavaScript bindings.
> 
> Unless you want to support other editors too.

We're talking about a solution that essentially monkey-patches VS Code to support tree-sitter syntax highlighting via an extension. Maybe the Venn diagram of editors that support LSP but _don't_ support tree-sitter highlighting is larger than I'm thinking, but for the scope of this particular problem it seems to me that the added complexity of managing a client/server IPC is not worth the trouble for this experiment, at least for an initial proof-of-concept.

**Comment by dberlin** [comment]
> Having a tree-sitter-to-LSP tool would absolutely be the best next step forward. It doesn't solve all of the problems, but it solves the biggest one.
> 
> @SNDST00M (and others) if you're interested
> in messing with the code to take advantage of the LSP let me know and I'll get you access to the [experimental tree sitter](https://github.com/jeff-hykin/experimental-tree-sitter).
> 
> Integrating the LSP likely wouldn't be too hard, however there is a sperate challenge. The tree sitter grammars are actually too powerful for the theming system. The grammars tell you everything about every token (which is awesome) but themes are way more dumb about what they color. There isn't a good generic way to map tree-sitter-information to the existing themes. Either all existing themes would break (no backward compatiblity), or there needs to be a custom converter for each language, in addition to a language parser. Once those values have been converted, you can then likely send them to the LSP.
> 
I mean, for starters, you only have the construct the mapping once.  

As for  automatically generating the mapping, having had to do similar types of migrations (regexp based systems to real parsing based systems) there are ways to do this that work pretty well.  You are right it is very hard to get rid of human intervention, but you can do a lot of the work for people.

For example, 
1. For the textmate colorizers that have real test sets, you can see what the textmate grammar colorizes, and what token sequences tree-sitter is mapping those strings to, and make statistical guesses as to what you should map to what. You can then let the user confirm it, and generate the mapping.
2. For textmate colorizes that don't, you can enumerate the language the regex parses, and parse it with tree-sitter, and do the same thing.
This will work, but requires more effort and will spend more CPU time.
Depending how far down the rabbit hole you want to go, you can make it very efficient at enumeration, because tree-sitter should be able to tell you what starting letters/etc might possibly change the next token (if any), and so you can skip enumerating large parts of the regex language that won't change anything.
3. You can also do the first idea dynamically - have the tree-sitter grammar run in the background on the files as people use vscode, see how the colorizer colored them vs tree-sitter parse tokens, and collect the stats and send them somewhere.  Use those stats of actual usage to generate the mapping.

**Comment by jeff-hykin** [comment]
> I mean, for starters, you only have the construct the mapping once.

Once per language, and it needs to be maintained as the parser and language itself change. **But yes**, not only is it not insurmountable, but Atom actually has already done this for every language that it supports with the tree sitter.

**Comment by jeff-hykin** [comment]
@dberlin 
> As for automatically generating the mapping, having had to do similar types of migrations (regexp based systems to real parsing based systems) there are ways to do this that work pretty well


Maybe you know of a system I don't, but myself and matter123 looked into this kind of automated system a bit (we already have a automated scope testing system for the C++ textmate library) and it was a bit harder than you might expect. If you do create or know of a system for this, please do share. I do work for Atom, and not only would this be great for tree sitters but it would also be very helpful for converting/updating themes.

**Comment by SNDST00M** [comment]
Tree-sitter and core grammars are available as an extension: https://github.com/EvgeniyPeshkov/syntax-highlighter

The extension can be installled from the Marketplace: https://marketplace.visualstudio.com/items?itemName=evgeniypeshkov.syntax-highlighter

@aeschli @fcurts @jeff-hykin @dannymcgee does this resolve the issue and/or present need(s)?

**Comment by jeff-hykin** [comment]
@SNDST00M it does not resolve this issue.

1. That extension must be forked or modified to add a new language. This is very different from the existing system where an extension can add a grammar for a new language or override the grammar for an existing language.
2. One of the main benefits of the tree sitter is efficiency. Right now that extension is running on top of the exiting textmate engine, making it significantly less efficient than using the textmate engine by itself.
3. The mapping from tree sitter scopes to semantic tokens is very very limited compared to what current themes can do with textmate grammars. C++ has over 400 different types of tokens, in contrast that extension supports about 8 types of tokens.
4. VS Code still uses textmate scopes to make internal decisions about things such as autosuggestions. Textmate has a sever limitation that it cannot look at multiple lines to make decisions. This means certain kinds of syntax just cannot be parsed. If VS Code uses textmate with a tree sitter overlay things like intelligent suggestions could still be broken even if the colors look correct.
5. Backwards compatibility is essentially non-existant. Existing themes are not aware of the semantic names added by the extension. And even if they are, many of the scopes they take advantage of are not available in the new system so they couldn't update themselves even if they wanted to.

The extension is great, but it's not the answer. It's a demonstration of what could be.

**Comment by dannymcgee** [comment]
> 4\. VS Code still uses textmate scopes to make internal decisions about things such as autosuggestions.

I could be wrong, but I don't think that's accurate. Completion suggestions are provided through their own [language feature hook](https://code.visualstudio.com/api/language-extensions/programmatic-language-features#show-code-completion-proposals), with the implementation left up to the individual language extension. AFAIK, the scopes defined by a TextMate grammar are used _only_ for the initial syntax highlighting pass and aren't even exposed to any other extension APIs. There is some sort of default completion provider that works in the absence of a language-specific one (e.g. in plain text docs), but as far as I can tell it just suggests whatever words already exist in that document.

I think the TextMate highlighting is pretty hard-wired into VS Code currently, so I don't think you could really turn it off without forking VS Code itself. There's a long-standing open issue about slow highlighting in [extreme edge cases](https://github.com/microsoft/vscode/issues/64681) where someone suggests moving the TM highlighting off the main thread, but I think that's about as much as you can hope for in terms of performance improvements over the current setup.

Honestly, I don't think efficiency is a good argument for tree-sitter in VS Code. VS Code's TextMate highlighter is already doing [incremental updates](https://code.visualstudio.com/blogs/2017/02/08/syntax-highlighting-optimizations) after the initial pass just like tree-sitter does, and I'm guessing (though I could be way off) that the "dumb" regex-based TM tokenizer would be faster than a tree-sitter CST parse, all else being equal.

The compelling arguments for tree-sitter vs TM grammars, IMO, are the semantic accuracy and (arguably) easier authorship. But to get the best semantic accuracy, you really need to parse not just the current file but also any external modules it's importing, which is always going to be slower than a simple line-by-line regex tokenizer.

**Comment by KamasamaK** [comment]
@dannymcgee TM scopes are used to determine token types (string, comment, other), which are used in many ways including in the `editor.quickSuggestions` setting. You are correct that they "aren't even exposed to any other extension APIs", which is probably why Jeff said "to make internal decisions".

And to add another thing TM scopes do, they are used to determine the (embedded) language that language snippets use.

**Comment by ebkgne** [comment]
> 2\. One of the main benefits of the tree sitter is efficiency. Right now that extension is running on top of the exiting textmate engine, making it significantly less efficient than using the textmate engine by itself.

@jeff-hykin I have installed that add-on to use instead of "c_cpp:Enhanced Colorization" , and the latency in comments colorization is hugely reduced in my case

Somebody said that our problem here ( https://github.com/microsoft/vscode/issues/64681 ) is caused by TM but then I think it might be something else  ?

**Comment by fcurts** [comment]
> The compelling arguments for tree-sitter vs TM grammars, IMO, are the semantic accuracy and (arguably) easier authorship. But to get the best semantic accuracy, you really need to parse not just the current file but also any external modules it's importing, which is always going to be slower than a simple line-by-line regex tokenizer.

Not the semantic accuracy, but the syntactic accuracy, which does *not* require looking at other files. It's impossible to do accurate syntactic syntax highlighting (and code folding, etc.) with TM grammars. Just take a look at the more than 700 (!) issues filed against the TypeScript TM grammar, the most complex regex spaghetti monster I've seen in my life. It's a never-ending "you fixed something here, but now it breaks when I add a newline there" nightmare. And that's *after* they added their own preprocessor to compose TM grammar fragments!

TM grammars need to die. It's a shame that after all these years, a leading editor such as VSCode offers nothing better.

**Comment by SNDST00M** [comment]
It's really annoying handling multi-line meta scopes like this: 

```matlab
classdef T15MultiLineClassdefHeader ... "it's a beautiful day to screw syntax up" - stinkmeaner
   < otherClass
```

**Comment by dannymcgee** [comment]
> Just take a look at the more than 700 (!) issues filed against the TypeScript TM grammar

That number literally includes _every bug ever filed_ against it. That repo is 5 years old. Filter by `label:bug -label:duplicate` and you get 87 actual bugs, 86 of which are closed. Less than 20 bugs per year on a grammar for a language that's under active development is really not the nightmare scenario you're making it out to be.

**Comment by fcurts** [comment]
> Filter by label:bug -label:duplicate and you get 87 actual bugs, 86 of which are closed.

I'm not sure "label:bug" is a reliable indicator. On the first issues page, only a single issue is marked with this label, yet almost every issue reports a highlighting bug.

Regardless, anyone who cares about correct syntax highlighting and has maintained a complex TM grammar (I have) will tell you that it's a nightmare and impossible to get right.

**Comment by gjsjohnmurray** [comment]
I was interested to discover recently that some of the VS Code core team are starting to leverage Tree-sitter

https://github.com/microsoft/vscode-anycode

**Comment by jasonwilliams** [comment]
@jeff-hykin from your experience, is this something that can be prototyped onto VSCode itself? or are there way too many changes? 

I've noticed so far the only attempts have been in the form of a plugin rather than a change to the core

**Comment by jeff-hykin** [comment]
@jasonwilliams there are way to many changes.   Actually, theres a linked issue somewhere in this thread from Alex that explains it in detail.


However Max Brunsfield (creator of the tree sitter) and the other Atom founders announced recently they're working on their own editor from scratch!

**Comment by michaelblyons** [comment]
https://zed.dev/

**Comment by jasonwilliams** [comment]
@jeff-hykin i have read Alex’s thread. I’m not convinced the issue is so much changing to tree sitter, it seems to be more moving tokenisation off the UI thread. Can tree sitter not be added by kept synchronous?

This is a hard problem but I don’t think it’s impossible. There may be some breaking changes along the way but it’s worth it.

Zedd looks interesting, I will keep an eye on it

**Comment by jeff-hykin** [comment]
@jasonwilliams true I guess that is the topic of that post.

Sadly though it is definitely not just a tokenizer that can be swapped out. Although semantic highlighting loosened this; from my understanding the theme system, internal functions, to code folding, to the syntax highlighting tags, along with many assumptions/optimizations are still tightly coupled to TextMate.

Atom was less inter-dependent internally because it was designed to be hackable, but it still had/has per-language manually-written conversion layers from tree sitter tokens to TextMate tokens because it's just so hard to fully do the tree sitter justice and not break everything.

**Comment by jasonwilliams** [comment]
It looks like for each language it goes off to fetch a tokenizator implementation (which happens to be the [TMTokenization class](https://github.com/microsoft/vscode/blob/main/src/vs/workbench/services/textMate/common/TMTokenization.ts#L11-L73) for now created [here](https://github.com/microsoft/vscode/blob/main/src/vs/workbench/services/textMate/browser/abstractTextMateService.ts#L259)

I wonder if as a start this can also fetch a treeSitterTokenization class which can then map its tree types back into textmate types so that things further upstream continue to work. It would need to implement [ITokenizationSupport](https://github.com/microsoft/vscode/blob/main/src/vs/editor/common/languages.ts#L277-L284)

My understanding is https://www.npmjs.com/package/web-tree-sitter would in all platforms web and desktop.

Tokenize is currently called line by line and expects a result in return.
The new tokenization class would need to return a tokenizationResult to the implementation of [EncodedTokenizationResult](https://github.com/microsoft/vscode/blob/main/src/vs/editor/common/languages.ts#L256-L272)

**Comment by meche-gh** [comment]
Is there real tangible data on the performance change from using Tree-sitter?

**Comment by jasonwilliams** [comment]
@meche-gh https://github.com/microsoft/vscode/pull/161479

**Comment by haikyuu** [comment]
Slightly tangent to this, I find tree-sitter to be interesting for more than syntax highlighting.
It's used in neovim, helix and other editors to power some very useful features:[ incremental selection, folding, indentation](https://github.com/nvim-treesitter/nvim-treesitter#available-modules) out of the box.

And most importantly, [external modules](https://github.com/nvim-treesitter/nvim-treesitter/wiki/Extra-modules-and-plugins) built around tree sitter are extremely useful.

I'd say we approach including tree-sitter in VSCode more holistically:
- How to make tree-sitter trees available for extensions: (attach the tree to the open file, get parent tree of current selection, and other utils ...) 
- How grammars are loaded: Currently, helix and nvim have a runtime directory that contains tree sitter grammars and queries. And it's picked up by the editor when it starts. 
- What utilities to bake into VSCode by default?

These decisions will likely impact the first inclusion of tree-sitter into VSCode, be it syntax highlighting or others.

I don't know if it's clear or not. But including tree-sitter into VSCode is a huge benefit because it makes it aware of the code and not treat it like text. It may start with syntax highlighting (which is a bit already solved by textmate grammars) but doesn't end there

**Comment by jasonwilliams** [comment]
Just giving my update and 2 cents.

@haikyuu those are some interesting thoughts, and I agree it’s a huge benefit all round.

I don’t agree about syntax highlighting being a solved problem because even though it “works” the performance is hitting its ceiling. I wrote about it here https://jason-williams.co.uk/posts/speeding-up-vscode-extensions-in-2022/ (see Tree Sitter section). If VSCode wants to stay competitive it will eventually need to migrate towards this in my opinion. Last time I looked at the performance of large files a lot of time was attributed to parsing.

I do agree with starting simple, but this will need to be in the core. I don’t want to see us go down a path of “everyone needs a tree sitter extension”, not that that’s what you were suggesting, but it would be good to see some roadmap for actually having it be the primary syntax system. 
My comment [above](https://github.com/microsoft/vscode/issues/50140#issuecomment-1086918405) looks into some first steps of adding it as a service then utilising it bit by bit, but still having the textmate system used primarily. This should provide at least _some_ migration path for  extensions to provide both textmate and treesitter trees.

I did look into branching of https://github.com/microsoft/vscode/pull/161479 but it’s a monumental effort as it touches so many parts of the code base. So it isn’t something I could take on alone, especially if the maintainers are already planning to work on this (we don’t know, they are quiet on this topic, although there’s still [positive](https://github.com/microsoft/vscode/issues/165506#issuecomment-1313890720) signals they’re interested in investigating).

There was concern over stability which may have been the reason progress in this area went quiet.

@alexdima did [raise](https://github.com/microsoft/vscode/pull/147648#issuecomment-1271499836) concerns around the ABI potentially changing causing extensions to break. 

Although I reached out to the Tree Sitter maintainers who declared the library to be [stable](https://github.com/microsoft/vscode/pull/161479#issuecomment-1368877912) and there shouldn’t be any backwards incompatible changes. 
Secondly Neovim, who have been using Tree Sitter for over 2 years, have only had forwards compatibility issues but not backwards. The former are more easily solved by having extensions build trees against a specific version before publishing.

**Comment by haikyuu** [comment]
@jasonwilliams I agree this should land into core for optimal experience. And the performance benefit is not to neglect (I am personally using neovim at the moment and everything feels way faster)

**Comment by M00TSDNS** [comment]
> There would need to be some mapping somehow from tree-sitter types back to TM Types, although `vscode-tree-sitter` does this so I'm guessing it's possible. If you map back there aren't a huge amount of changes, you could probably stop here and change things further upstream at a later point.

If the Tree-Sitter community wants to scale to existing themes, they need to plan such a mapping ahead of time and standardise it the way that [Sublime and Textmate did](https://github.com/microsoft/vscode/issues/126957#issuecomment-866581927), and also the way Microsoft began to do with the LSP token format a couple months in.

**Comment by M00TSDNS** [comment]
Nvm, if the mapping is done by the grammar owner, that would be a small portion of the current effort needed to maintain Textmate grammars.. Would suck more if there was no TM grammar but even then the mapping would only be painful once

**Comment by MixusMinimax** [comment]
> I fully agree with you that TextMate grammars are challenging to implement and have limitations, but it's always a lot of work to create and maintain a grammar. That will not be different with Tree-Sitter.

While I agree with the fact that tree-sitter grammars also have their difficulties, the difference is that many of us need to define a tree-sitter grammar anyway, since you can use it for other things, like an LSP server, or a compiler. A textmate grammar would have to be maintained _in parallel_ with whatever other parser generator you're using for other components.

**Comment by codethief** [comment]
https://github.com/microsoft/vscode/pull/161479#issuecomment-1761967068 sounds like tree-sitter support is not going to happen(?) :\

**Comment by texastoland** [comment]
The team hasn't been active in any linked issues, hasn't publicly expressed intent to change direction, and arguably has competing interests with unifying VS Code syntax definitions with either VS and/or Monaco/Monarch. I consider this issue closed in practice.

**Comment by texastoland** [comment]
> so ... where do we go to ask ... "either VS and/or Monaco/Monarch" to adopt treesitter :)

#### Visual Studio

- Issues: https://developercommunity.visualstudio.com/VisualStudio/suggest
- Official extension: https://github.com/microsoft/vscode-anycode

#### Monarch (Monaco)

- Issues: https://github.com/microsoft/monaco-editor
- Documentation
  - https://microsoft.github.io/monaco-editor/monarch.html
  - https://microsoft.github.io/monaco-editor/docs.html#interfaces/languages.LanguageConfiguration.html
- Context: https://github.com/microsoft/vscode/issues/216#issuecomment-222501886

**Comment by heartacker** [comment]
https://github.com/microsoft/vscode/issues/206739 🛩️ 

## Code Editor
 Explore using the new EditContext API https://github.com/microsoft/vscode/issues/204371 @hediet
 💪 Explore hover enriching https://github.com/microsoft/vscode/issues/195394 @aiday-mar @hediet
 🔴 Explore tree-sitter parser ecosystem @alexr00

**Comment by texastoland** [comment]
> #206739 🛩️

> > [Explore tree-sitter parser ecosystem](https://arc.net/l/quote/bejyuqqe)

Well dam 🙊

**Comment by michaelblyons** [comment]
https://github.com/microsoft/vscode/issues/207416 for those who missed it.

**Comment by Milo123459** [comment]
That issue is now locked. Has any further exploration occurred?

**Comment by RedCMD** [comment]
yes
you can follow along at https://github.com/microsoft/vscode/issues/210475#issuecomment-2325306943

**Comment by RedCMD** [comment]
@mantou132 are you asking how to create a grammar to syntax highlight a file like that image?
https://code.visualstudio.com/api/language-extensions/syntax-highlight-guide
https://marketplace.visualstudio.com/items?itemName=RedCMD.tmlanguage-syntax-highlighter
